ceno_zkvm/src/scheme/prover.rs

```diff
@@ -103,20 +103,47 @@ impl<E: ExtensionField, PCS: PolynomialCommitmentScheme<E>> ZKVMProver<E, PCS> {
                 circuit_name = circuit_name,
                 profiling_2 = true
             );
-            let witness = match num_instances {
-                0 => vec![],
+            let num_witin = self
+                .pk
+                .circuit_pks
+                .get(&circuit_name)
+                .unwrap()
+                .get_cs()
+                .num_witin;
+            let (witness, structural_witness) = witness.split_at(num_witin as usize);
```

Comment:
`split_at` is a bit complex. We can do more simple way by split_off on result vector

```diff
diff --git a/ceno_zkvm/src/scheme/prover.rs b/ceno_zkvm/src/scheme/prover.rs
index 0defa093..0bdc40b0 100644
--- a/ceno_zkvm/src/scheme/prover.rs
+++ b/ceno_zkvm/src/scheme/prover.rs
@@ -110,18 +110,16 @@ impl<E: ExtensionField, PCS: PolynomialCommitmentScheme<E>> ZKVMProver<E, PCS> {
                 .unwrap()
                 .get_cs()
                 .num_witin;
-            let (witness, structural_witness) = witness.split_at(num_witin as usize);
-
             let (witness, structural_witness) = match num_instances {
                 0 => (vec![], vec![]),
                 _ => {
-                    let witness = witness.into_mles();
+                    let mut witness = witness.into_mles();
+                    let structural_witness = witness.split_off(num_witin as usize);
                     commitments.insert(
                         circuit_name.clone(),
                         PCS::batch_commit_and_write(&self.pk.pp, &witness, &mut transcript)
                             .map_err(ZKVMError::PCSError)?,
                     );
-                    let structural_witness = structural_witness.into_mles();

                     (witness, structural_witness)
                 }
```

and we can remove `split_at` implementation